src/main/java/org/wildfly/security/auth/realm/token/TokenSecurityRealm.java

```diff
@@ -44,6 +45,8 @@
 
     private final TokenValidator strategy;
     private final String principalClaimName;
+    /** A function that extracts the principal name from the set of token claims. Takes precedence over {@linkplain #principalClaimName} */
+    private final Function<Attributes, String> principalClaimNameFunc;
```

Comment:
Maybe it should be a `Function<Attributes, Principal>`?

Reply:
I can see adding that instead.